### PR TITLE
Simplify Domain.join

### PR DESCRIPTION
**Pull request by kayceesrk**
This PR simplifies `Domain.join` by using mutex and condition variables, and avoids the more complicated logic used currently on `trunk`. The termination mutex is no longer shared with the runtime which makes the reasoning simpler. 

The PR also introduces a user-visible change -- joining more than once on a domain no longer raises an exception. The proposed semantics for `Domain.join` returns the result of `Domain.spawn` if the computation ran to completion normally and raises the same exception if the computation had raised an exception, irrespective of the number of times `Domain.join` is called on a domain. It was unclear to me why the current "exactly-once" restriction was introduced in the first place. I assume that it may have to do with the lack of Mutex and Condition modules in the previous implementations of multicore.

Alternative semantics for the handling of uncaught exceptions is discussed in #11074. This PR aims to only introduce the minimal change in semantics.

### CODE REVIEW

**Comment by gasche** [comment]
I agree that this (having the lock purely on the OCaml side) is simpler than having specific support in the runtime.

A question about the new implementation: you are using a condition variable to signal termination, but unless I am missing something you could follow the same approach as the runtime code would do, which is to have just a lock that is locked by the domain while it is running, and that joiners can try to lock. Why is it better to use a condition variable here?

(For STW I asked the same section, and the answer was that a different domain could terminate STW-exit, but this does not apply here.)

**Comment by kayceesrk** [comment]
This is an idiomatic case for the use of condition variables. The use of condition variables avoids spin waiting.

In the solution without condition variables, the `term_mutex` has to be locked by the child domain and has to be held throughout its execution. For this to happen, the `term_mutex` has to be locked by the child after it starts running (and not by the parent domain to avoid syserror on unlocking by the child). However, the parent may try to join before the child has had a chance to lock the `term_mutex`. In this case, the parent (or indeed, other domains trying to join the child domain) will have to spin wait until the child gets a chance to lock the `term_mutex`. This is why the current implementation in trunk has a spin-wait loop. 

With the condition variable `term_condition`, the other domains may park themselves on the condition variable rather than spin waiting. Condition variables are much nicer than hand-written spin-wait loops.

**Review by gasche — APPROVED** [review]
Thanks for the extra explanation. I think this is good to go if the CI agrees.

**Comment by kayceesrk** [comment]
The tests `finaliser2.ml` and `beat.ml` are failing in the CI, and both are known to fail intermittently. I have proposed a fix for `finaliser2.ml` failure in #11201.

**Review by sadiqj — APPROVED** [review]
LGTM. It's nice we've moved code out of the runtime and in to Domain.

I've kicked the CI a few times to rerun failed jobs and `beat.ml` seems to be failing pretty consistently now though.

**Comment by gasche** [comment]
According to github, all the tests are passing now, so I'm moving to merge now. There is something to investigate in `beat.ml`, but I believe it is not related to the present PR (it is a single-domain test for Thread).

**Comment by gadmm** [comment]
I am not keen on the idea that the same exception can be raised several times. It never makes sense to handle the same exception several times, and I cannot think of a situation where this would not reveal a bug in the program's error handling.

**Comment by kayceesrk** [comment]
@gadmm we should continue discussion of the uncaught exception semantics on #11074.

**Comment by xavierleroy** [comment]
#11208 shows a serious issue with this reimplementation of `Domain.join`: even after `join` returned, the thread can still have arbitrary amounts of OCaml code to run before actually terminating.  

This makes the idiom "join all domains before exiting the whole process" useless: some domains will still be running when the whole process exits.  This prevents proper finalization in domains and can cause general protection faults under Windows.

I think we need a different implementation of `Domain.join` that builds on `pthread_join` to make sure the domain is completely finished.  The one-place buffer (with mutex and condition variable) in OCaml can still be used to pass the return value of the domain.  But "return value is available" is not a proper substitute for "the domain has terminated".

**Comment by gasche** [comment]
I'm not sure I understand the issue. By "arbitrary amounts of OCaml code", you mean finalisers? With the previous design, finalisers would not be called on the finishing domain, but adopted by some other domain before `join` returns, so they were not "code to run before actually terminating" -- they would run after termination, like today. The difference described in #11208 is whether finalisers are adopted before or after `join` succeeds, not where they run.

On the other hand, maybe there are other issues with joining purely on the OCaml side; for example could a signal handler still run on the domain after it signalled its own termination? (Can we avoid this by unregistering signal handlers on the OCaml side? Sounds a bit fragile.) If we want to guarantee that no OCaml code runs on the domain after joining, we probably have more thinking to do indeed.

**Comment by kayceesrk** [comment]
> even after join returned, the thread can still have arbitrary amounts of OCaml code to run before actually terminating.

I'm trying to understand this comment better. The last thing that a domain does is signal the condition variable and release the termination mutex:

https://github.com/ocaml/ocaml/blob/f27d671b23f5246d37d91571eeccb802d5399a0b/stdlib/domain.ml#L209-L210

The runtime function `domain_terminate` hands over the finaliser callbacks, and does not run the finaliser callbacks on this domain. Is there something else that I am missing?

**Comment by gasche** [comment]
The following paragraph of the `pthread_join` manpage on my system sounds mildly worrying:

> Faiilure to join with a thread that is joinable (i.e., one that  is  not
   detached),  produces  a  "zombie thread".  Avoid doing this, since each
   zombie thread consumes some system resources, and  when  enough  zombie
   threads  have  accumulated, it will no longer be possible to create new
   threads (or processes).

It's only "mildly worrying" as pthread can handle thousands of threads easily, while we are not expected thousands of domains to be created, but it does suggest that calling `pthread_join` would be a good idea.

The manpage also says that it is undefined to have several calls to `pthread_join` on the same thread. This may explain why the `Domain.join` API was chosen to have this same interface (a single joining domain).

**Comment by xavierleroy** [comment]
OK, I was under the wrong impression that finalizers could run in a domain after its main OCaml function (the one in domain.ml) has returned.  

> The last thing that a domain does is signal the condition variable and release the termination mutex:

The point I'm trying to make is that this is the last thing the main OCaml function for the domain does, but the domain keeps running a bit longer in C, doing whatever resource finalization it has to do, before actually terminating.

So, with the current implementation, `Domain.join` can return while the domain is still running (as far as the OS is concerned), and I'm worried this can have bad consequences.

**Comment by xavierleroy** [comment]
> It's only "mildly worrying" as pthread can handle thousands of threads easily, while we are not expected thousands of domains to be created, but it does suggest that calling pthread_join would be a good idea.

Currently, OCaml creates threads in detached mode, where explicit joining is not needed to reclaim resources.  But joining all threads/domains that have been created is a good programming discipline.

As to joining only once, this can be worked around easily if we want.

**Comment by kayceesrk** [comment]
> where explicit joining is not needed to reclaim resources

Joining a detached thread seems to be undefined behaviour [1]. If threads are neither detached nor joined, are we only leaking resources or worse? 

[1] https://wiki.sei.cmu.edu/confluence/display/c/CON39-C.+Do+not+join+or+detach+a+thread+that+was+previously+joined+or+detached

**Comment by xavierleroy** [comment]
> If threads are neither detached nor joined, are we only leaking resources or worse?

Leaking resources, as far as I know.  Like with an unclosed file descriptor or a zombie process.

**Comment by gadmm** [comment]
> On the other hand, maybe there are other issues with joining purely on the OCaml side; for example could a signal handler still run on the domain after it signalled its own termination? (Can we avoid this by unregistering signal handlers on the OCaml side? Sounds a bit fragile.) If we want to guarantee that no OCaml code runs on the domain after joining, we probably have more thinking to do indeed.

Well yes, and so it can call finalisers for the same reasons. And that is a problem since you now have a race where it returns as if a value was produced, but an async exception is secretly ignored. (It is fine to not run an async callback that is going to produce an exception; it is not fine to silently ignore the exception once it has been produced.)

I am not sure what to think of the joining being done on the OCaml side, but if you want to keep this design, one could "turn off" async callbacks before termination. By that I mean disabling the automatic running of callbacks at safe points while keeping them in the queue (for running them later, if necessary after adoption by another domain). For lack of a better word, let us call this putting a "veil" on async callbacks. Well I think that a "veil" should already be on before the at_exit handlers are called, and I conjecture (but it's just a conjecture, I have not looked at it too closely) that "veils" are not too difficult to implement, once my PR on async actions is merged.

**Comment by gadmm** [comment]
An issue with joining on the OCaml side is that a domain is a resource than you want to be cleaned-up promptly. Indeed if the program creates domains again after joining others, then you want to be able to reason that you are not going to exceed `Max_domains` running in parallel.

But you still need a "veil" for at_exit handlers (including Stdlib.at_exit), but this is another issue.

**Comment by kayceesrk** [comment]
I've been thinking about a design that would encompass the various points raised. In particular, we want `Domain.join` to really signify that the target domain has completed its execution. In particular,

* No OCaml or C code should run on the target domain `d` when `Domain.join d` returns. So no signal handlers, callbacks, etc. 
* The program doesn't spuriously exceed `Max_domains` limit since the domain slots may still be active 

The previous implementation of `Domain.join`, which was modified by this PR, almost did the right thing. The waiting domains are woken up only after `domain_terminate` function was called. So neither the domain itself nor the backup thread will participate in the GC. Since the domain also removes itself from the participating set (see `domain.c:remove_from_stw_domains`), we will not spuriously exceed `Max_domains` limit. The earlier implementation already ensures that no callbacks are run on the terminating domain before the waiting domains are woken up.

The only guarantee that the previous design does not fulfil is that there is some C code that runs after waiting domains are signalled. 

https://github.com/ocaml/ocaml/blob/949e2626c22b5fe8159cb29d47207d5621eabf90/runtime/domain.c#L1106-L1114

There is a call to `free_domain_ml_values` which I believe can be moved prior to releasing the mutex. We may also block signals on this pthread before releasing the mutex to ensure that signal handlers don't run.

## Building on `pthread_join`

I thought about the alternative design of building on top of `pthread_join` for `Domain.join`. The resultant API is more complex for the end-user. 

Currently, the thread for the domain is created in detached mode. Hence, there is no need to `pthread_join` on this thread for cleaning up resources. If we were to build on top of `pthread_join`, then the thread for the domain cannot be detached just after creation as joining a detached thread is undefined behaviour. We will have to ask the users to join every domain that they have created. However, this is not possible on certain platforms as @dbuenzli pointed out as blocking the main thread is forbidden in `macos` and `ios`. So we will have to also expose `Domain.detach` functionality. Despite this, the users may neither detach nor join threads, which will lead to resource leaks. 

## Proposal

Hence, I am of the opinion that it may be best to revert this PR which retains the simple API and nice resource cleanup behaviour. On top of the revert, we will strengthen the terminate function to block signals and release the mutex as the last step.

**Comment by gasche** [comment]
(Minor note: For me it's okay if some C code still has to run on the domain by the time it becomes "joinable", but ideally this would be cleanup/runtime code that is not observable by the user.)

**Comment by kayceesrk** [comment]
If we choose to revert this PR, then #11208 is not necessary. 

@xavierleroy what do you think about the proposal in https://github.com/ocaml/ocaml/pull/11196#issuecomment-1103546040?

**Comment by xavierleroy** [comment]
I'm catching up with the discussion.  I agree that rebuilding Domain.join on top of pthread_join is a big change of API and would require a lot more thinking and experimentation.  I'm in favor of @kayceesrk 's proposal: revert this PR and block signals before signalling termination.  This should be enough to make Domain.join work like we expect it to.  

At some point we'll need to understand the Windows situation better (i.e. why exiting the program while domains or threads are still running often causes a crash?), but it's an old problem, so it can wait.

**Comment by kayceesrk** [comment]
I've opened #11214 which addresses the issues being discussed.

**Pull request merged**
